File: sdk/python/src/dograh_sdk/_validation.py
"""Client-side validation of node data against a fetched spec.

Intentionally lightweight: we catch the hallucinations that matter (typo'd
field names, missing required fields, obvious scalar-type mismatches) and
leave rigorous coercion to the backend's Pydantic validators, which run at
save time. The tradeoff: the SDK fails fast on mistakes an LLM makes, and
the backend remains the single authority on wire-format correctness.
"""

from __future__ import annotations

from typing import Any

from .errors import ValidationError

# Map PropertyType → primitive Python type(s) we check at call site.
# `None` means "skip scalar-type check" (compound types, refs, JSON, etc.).
_SCALAR_TYPES: dict[str, tuple[type, ...] | None] = {
    "string": (str,),
    "number": (int, float),
    "boolean": (bool,),
    "options": None,  # value-in-options handled separately
    "multi_options": None,
    "fixed_collection": (list,),
    "json": None,  # any JSON-serializable
    "tool_refs": (list,),
    "document_refs": (list,),
    "recording_ref": (str,),
    "credential_ref": (str,),
    "mention_textarea": (str,),
    "url": (str,),
}


def _with_hint(prop: dict[str, Any], message: str) -> str:
    """Append `prop.llm_hint` to an error message when set.

    Surfacing the hint inside validation errors lets an LLM author
    self-correct on retry — it sees the catalog reference or value-shape
    guidance inline with the failure.
    """
    hint = prop.get("llm_hint")
    if hint:
        return f"{message}\n  Hint: {hint}"
    return message


def _check_scalar(prop: dict[str, Any], value: Any) -> None:
    # None is always allowed (missing value handled by required check).
    if value is None:
        return
    allowed = _SCALAR_TYPES.get(prop["type"])
    if allowed is None:
        return
    # Booleans ARE ints in Python, so exclude accidentally-matching bools.
    if bool in allowed and not (int in allowed or float in allowed):
        if not isinstance(value, bool):
            raise ValidationError(
                _with_hint(
                    prop,
                    f"{prop['name']}: expected boolean, got {type(value).__name__}",
                )
            )
        return
    if not isinstance(value, allowed):
        raise ValidationError(
            _with_hint(
                prop,
                f"{prop['name']}: expected {prop['type']}, "
                f"got {type(value).__name__}",
            )
        )


def _check_options(prop: dict[str, Any], value: Any) -> None:
    if value is None:
        return
    allowed = {o["value"] for o in prop.get("options") or []}
    if not allowed:
        return
    if prop["type"] == "multi_options":
        if not isinstance(value, list):
            raise ValidationError(
                _with_hint(
                    prop,
                    f"{prop['name']}: expected list, got {type(value).__name__}",
                )
            )
        bad = [v for v in value if v not in allowed]
        if bad:
            raise ValidationError(
                _with_hint(
                    prop,
                    f"{prop['name']}: values {bad} not in allowed {sorted(allowed)}",
                )
            )
    else:  # 'options'
        if value not in allowed:
            raise ValidationError(
                _with_hint(
                    prop,
                    f"{prop['name']}: {value!r} not in allowed {sorted(allowed)}",
                )
            )


def validate_node_data(
    spec: dict[str, Any],
    kwargs: dict[str, Any],
) -> dict[str, Any]:
    """Validate LLM-supplied kwargs against the node spec. Returns the data
    dict to embed in the wire format, with defaults applied.

    Raises:
        ValidationError if kwargs contain unknown fields, omit required
        fields, or carry obvious type mismatches.
    """
    declared = {p["name"]: p for p in spec["properties"]}

    # Unknown field names — the most common LLM hallucination.
    unknown = set(kwargs) - set(declared)
    if unknown:
        raise ValidationError(
            f"{spec['name']}: unknown field(s) {sorted(unknown)}. "
            f"Allowed: {sorted(declared)}"
        )

    # Per-property validation
    data: dict[str, Any] = {}
    for name, prop in declared.items():
        if name in kwargs:
            value = kwargs[name]
        elif prop.get("default") is not None:
            value = prop["default"]
        else:
            value = None

        # Scalar / collection shape
        if prop["type"] in ("options", "multi_options"):
            _check_options(prop, value)
        else:
            _check_scalar(prop, value)

        # Nested fixed_collection rows — validate each row as a sub-spec.
        if prop["type"] == "fixed_collection" and isinstance(value, list):
            sub_spec = {"name": f"{spec['name']}.{name}", "properties": prop.get("properties") or []}
            data[name] = [validate_node_data(sub_spec, row) for row in value]
            continue

        if value is not None:
            data[name] = value

    # Required check — must be set AND non-empty for strings.
    for name, prop in declared.items():
        if not prop.get("required"):
            continue
        val = data.get(name)
        if val is None or (isinstance(val, str) and val.strip() == ""):
            raise ValidationError(
                _with_hint(
                    prop,
                    f"{spec['name']}: required field missing: {name}",
                )
            )

    return data

File: sdk/typescript/src/validation.ts
// Client-side validation of node data against a fetched spec. Mirrors
// `sdk/python/src/dograh_sdk/_validation.py` byte-for-byte where possible
// so the two SDKs raise identical error messages for identical bad input.
//
// Intentionally lightweight: catch typos / missing required / obvious
// scalar mismatches at the call site; leave rigorous coercion to the
// backend Pydantic validators at save time.

import { ValidationError } from "./errors.js";
import type { NodeSpec, PropertySpec } from "./types.js";

// PropertyType → expected JS typeof values (after accounting for `null`
// and arrays). `null` here means "skip scalar-type check" (compound
// types, refs, JSON, etc.).
const SCALAR_TYPES: Record<string, ReadonlyArray<string> | null> = {
    string: ["string"],
    number: ["number"],
    boolean: ["boolean"],
    options: null,
    multi_options: null,
    fixed_collection: ["array"],
    json: null,
    tool_refs: ["array"],
    document_refs: ["array"],
    recording_ref: ["string"],
    credential_ref: ["string"],
    mention_textarea: ["string"],
    url: ["string"],
};

function jsTypeOf(value: unknown): string {
    if (value === null) return "null";
    if (Array.isArray(value)) return "array";
    return typeof value;
}

function withHint(prop: PropertySpec, message: string): string {
    return prop.llm_hint ? `${message}\n  Hint: ${prop.llm_hint}` : message;
}

function checkScalar(prop: PropertySpec, value: unknown): void {
    if (value === undefined || value === null) return;
    const allowed = SCALAR_TYPES[prop.type];
    if (!allowed) return;
    const got = jsTypeOf(value);
    if (!allowed.includes(got)) {
        throw new ValidationError(
            withHint(prop, `${prop.name}: expected ${prop.type}, got ${got}`),
        );
    }
}

function checkOptions(prop: PropertySpec, value: unknown): void {
    if (value === undefined || value === null) return;
    const allowed = new Set((prop.options ?? []).map((o) => o.value));
    if (allowed.size === 0) return;
    if (prop.type === "multi_options") {
        if (!Array.isArray(value)) {
            throw new ValidationError(
                withHint(
                    prop,
                    `${prop.name}: expected list, got ${jsTypeOf(value)}`,
                ),
            );
        }
        const bad = value.filter(
            (v) => !allowed.has(v as string | number | boolean),
        );
        if (bad.length > 0) {
            throw new ValidationError(
                withHint(
                    prop,
                    `${prop.name}: values ${JSON.stringify(bad)} not in allowed ${JSON.stringify(
                        [...allowed].sort(),
                    )}`,
                ),
            );
        }
    } else if (!allowed.has(value as string | number | boolean)) {
        throw new ValidationError(
            withHint(
                prop,
                `${prop.name}: ${JSON.stringify(value)} not in allowed ${JSON.stringify(
                    [...allowed].sort(),
                )}`,
            ),
        );
    }
}

export function validateNodeData(
    spec: NodeSpec | { name: string; properties: PropertySpec[] },
    kwargs: Record<string, unknown>,
): Record<string, unknown> {
    const declared = new Map(spec.properties.map((p) => [p.name, p]));

    // Unknown field names — the most common LLM hallucination.
    const unknown = Object.keys(kwargs).filter((k) => !declared.has(k));
    if (unknown.length > 0) {
        throw new ValidationError(
            `${spec.name}: unknown field(s) ${JSON.stringify(unknown.sort())}. ` +
                `Allowed: ${JSON.stringify([...declared.keys()].sort())}`,
        );
    }

    const data: Record<string, unknown> = {};
    for (const [name, prop] of declared) {
        let value: unknown;
        if (name in kwargs) {
            value = kwargs[name];
        } else if (prop.default !== undefined && prop.default !== null) {
            value = prop.default;
        } else {
            value = undefined;
        }

        if (prop.type === "options" || prop.type === "multi_options") {
            checkOptions(prop, value);
        } else {
            checkScalar(prop, value);
        }

        // Nested fixed_collection rows — validate each row as a sub-spec.
        if (prop.type === "fixed_collection" && Array.isArray(value)) {
            const subSpec = {
                name: `${spec.name}.${name}`,
                properties: prop.properties ?? [],
            };
            data[name] = value.map((row) =>
                validateNodeData(subSpec, row as Record<string, unknown>),
            );
            continue;
        }

        if (value !== undefined) data[name] = value;
    }

    // Required check — must be set AND non-empty for strings.
    for (const [name, prop] of declared) {
        if (!prop.required) continue;
        const val = data[name];
        if (
            val === undefined ||
            val === null ||
            (typeof val === "string" && val.trim() === "")
        ) {
            throw new ValidationError(
                withHint(
                    prop,
                    `${spec.name}: required field missing: ${name}`,
                ),
            );
        }
    }

    return data;
}
